### DISCUSSION TITLE
What is the list of all existing Bril interpreters?

### DISCUSSION THREAD

**Discussion by rcplane** [sampsyo/bril#297] [Q&A] [answered]
I am interested in the question of what is the fastest Bril interpreter.
In order to form a leaderboard, we would need

1. a list of interpreters that exist
2. specify a benchmark suite and running conditions (runtime, end-to-end including compilation time)
3. a standardized platform target such as an x86 virtualized docker container based on polyglot for github actions submission

Starting from the top, what are all existing Bril interpreters? Please do contribute more to this list, especially if the interpreter in question has a shot at fastest. If particular benchmarks have performance interest, do note them.

- [brili.ts](https://github.com/sampsyo/bril/blob/main/bril-ts/bril.ts) [doc](https://capra.cs.cornell.edu/bril/tools/interp.html)
- [bril-rs](https://github.com/sampsyo/bril/blob/main/bril-rs/brillvm/src/main.rs) [doc](https://github.com/sampsyo/bril/tree/main/bril-rs/brillvm)
- [brilift](https://github.com/sampsyo/bril/blob/main/brilift/src/main.rs)
- [brilc](https://github.com/sampsyo/bril/blob/main/bril-llvm/brilc)
- [bril-ocaml](https://github.com/sampsyo/bril/blob/main/bril-ocaml/lib/bril.ml)
- [bril-hs](https://github.com/zachary-kent/bril/blob/ssa/bril-hs/app/Main.hs)

**Comment by Pat-Lafon** [comment]
Have you seen https://github.com/sampsyo/bril/pull/213 https://github.com/sampsyo/bril/pull/234? There is also https://github.com/sampsyo/cs6120/issues/304 floating around.

I would also note something like: https://github.com/sampsyo/bril/pull/221. Creating `benchmarks/long` with the purpose of stressing interpreters(as opposed to programs which are amenable to compiler optimizations).

**Comment by sampsyo** [comment, marked as answer]
There are two in-tree interpreters:

* [the slow reference interpreter in TypeScript](https://capra.cs.cornell.edu/bril/tools/interp.html)
* [the fast interpreter in Rust](https://capra.cs.cornell.edu/bril/tools/brilirs.html)

There are three in-tree compilers from Bril to native executables:

* [Brilift](https://capra.cs.cornell.edu/bril/tools/brilift.html), using Cranelift
* [Brillvm](https://github.com/sampsyo/bril/tree/main/bril-rs/brillvm), using LLVM via some Rust bindings
* [bril-llvm](https://github.com/sampsyo/bril/tree/main/bril-llvm), using LLVM via generating text IR from Python

I don't believe the OCaml library includes an interpreter.

Some things that are not in tree that could be worth trying, but for which we can't vouch for the "freshness":

* a [bytecode interpeter](https://www.cs.cornell.edu/courses/cs6120/2022sp/blog/fast-brili/) (I would still love to merge this into the tree if you want, @charles-rs & @susan-garry)
* a [compiler to C source code](https://github.com/xu3kev/bril2c)
* a [compiler to Java bytecode](https://github.com/Neroysq/bril2jb)

**Reply by charles-rs** [reply]
Give me like a week or so to split off of the byte code stuff from the armv8 stuff then merge it in!

**Reply by sampsyo** [reply]
Awesome!

**Reply by rcplane** [reply]
Hi @Pat-Lafon , might you be able to recommend a version of LLVM to try with [Brillvm](https://github.com/sampsyo/bril/tree/main/bril-rs/brillvm) ?
I've been trying `make rt` with LLVM 17 and seeing
```
error: could not find native static library `Polly`, perhaps an -L flag is missing?
error: could not compile `llvm-sys` (lib) due to previous error
```

**Reply by Pat-Lafon** [reply]
Inkwell is working on support for llvm 17 https://github.com/TheDan64/inkwell/pull/448. Brillvm currently should work with llvm16 as specified in the CI. Maybe I should note that.
